File: AlgorithmsPackPythonSolutions/Course Schedule II.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        graph = [[] for _ in range(numCourses)]
        incomingEdges = [0] * numCourses

        for prerequisite in prerequisites:
            fromCourse, toCourse = prerequisite
            graph[toCourse].append(fromCourse)
            incomingEdges[fromCourse] += 1

        queue = deque()
        for i in range(numCourses):
            if incomingEdges[i] == 0:
                queue.append(i)

        nodesFound = 0
        result = []
        while queue:
            course = queue.popleft()
            nodesFound += 1
            result.append(course)

            for neighbor in graph[course]:
                incomingEdges[neighbor] -= 1
                if incomingEdges[neighbor] == 0:
                    queue.append(neighbor)

        if nodesFound != numCourses:
            result.clear()
        
        return result

File: AlgorithmsPackCPP/Course Schedule II.cpp
class Solution {
    typedef vector<vector<int>>GRAPH;
    
    void addEdge(GRAPH&graph, int from, int to){
        graph[from].push_back(to);
    }
public:
    vector<int>topSort(GRAPH&graph){
        int nodes = graph.size();
        vector<int>incomingEdges(nodes);
        for(int i = 0; i < nodes; i++){
            for(int j:graph[i])
                incomingEdges[j]++;
        }
        queue<int>q;
        for(int i = 0; i < nodes; i++){
            if(incomingEdges[i]==0)
                q.push(i);
        }
        int nodesFound = 0;
        vector<int>result;
        while(!q.empty()){
            int i = q.front();
            q.pop();
            nodesFound++;
            result.push_back(i);
            
            for(int j:graph[i]){
                if(--incomingEdges[j]==0)
                    q.push(j);
            }
        }
        if(nodesFound!=nodes)
            result.clear();
        return result;
    }
    
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        GRAPH graph(numCourses);
        for(auto&pre:prerequisites){
            addEdge(graph,pre[1],pre[0]);
        }
        
        return topSort(graph);
    }
};
